### ISSUE TITLE
Installing inside a monorepo causes "Cannot find module" errors

### ISSUE THREAD

**Issue by EvHaus** [wxt-dev/wxt#457]
### Describe the bug

I'm testing WXT for the first time. I create a new package inside my Yarn monorepo and ran the `pnpx wxt@latest init wxt-test` command. Unfortunately running `yarn install` fails from here:

```
error /my-monorepo/node_modules/wxt-starter: Command failed.
Exit code: 1
Command: wxt prepare
Arguments:
Directory: /my-monorepo/node_modules/wxt-starter
Output:
WXT 0.16.8
ℹ Generating types...

 ERROR  Cannot find module '/my-monorepo/packages/wxt-test/node_modules/wxt/dist/virtual/mock-browser.js'
Require stack:
- /my-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js

  Require stack:
  - /my-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js
  at Module._resolveFilename (node:internal/modules/cjs/loader:1072:15)
  at Function.resolve (node:internal/modules/helpers:127:19)
  at _resolve (/my-monorepo/node_modules/jiti/dist/jiti.js:1:251148)
  at jiti (/my-monorepo/node_modules/jiti/dist/jiti.js:1:253746)
  at /my-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js:33:44
  at evalModule (/my-monorepo/node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (/my-monorepo/node_modules/jiti/dist/jiti.js:1:254371)
  at /my-monorepo/node_modules/wxt/dist/browser.js:23:29
  at evalModule (/my-monorepo/node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (/my-monorepo/node_modules/jiti/dist/jiti.js:1:254371)

✖ Command failed after 417 ms

 ERROR  Failed to load entrypoint: entrypoints/background.ts

  at importEntrypointFile (/my-monorepo/node_modules/wxt/dist/cli.js:2027:13)
  at async getBackgroundEntrypoint (/my-monorepo/node_modules/wxt/dist/cli.js:505:27)
  at async /my-monorepo/node_modules/wxt/dist/cli.js:274:18
  at async Promise.all (index 0)
  at async findEntrypoints (/my-monorepo/node_modules/wxt/dist/cli.js:264:23)
  at async prepare (/my-monorepo/node_modules/wxt/dist/cli.js:3448:23)
```

I suspect this is happening because some WXT script is trying to run `/my-monorepo/packages/wxt-test/node_modules/wxt/dist/virtual/mock-browser.js` but that module actually gets installed 2 levels above in `/my-monorepo/node_modules/wxt/dist/virtual/mock-browser.js`

### To Reproduce

Steps to reproduce the bug using the reproduction:

1. Setup a barebones [yarn](https://yarnpkg.com/features/workspaces) or [pnpm](https://pnpm.io/workspaces) monorepo
2. `cd packages/`
3. `pnpx wxt@latest init wxt-test`
4. `cd wxt-test`
5. `yarn`

### Expected behavior

WXT is able to install the project and run the necessary scripts.

### Screenshots

N/A

### Environment

```
  System:
    OS: macOS 14.2.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 86.36 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.3.1 - ~/.nvm/versions/node/v20.3.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v20.3.1/bin/yarn
    npm: 9.8.0 - ~/.nvm/versions/node/v20.3.1/bin/npm
  Browsers:
    Chrome: 121.0.6167.184
    Chrome Canary: 123.0.6304.0
    Edge: 121.0.2277.128
    Firefox: 121.0.1
    Safari: 17.2.1
```

### Additional context

N/A

**Comment by aklinker1** [comment]
I have several monorepos, and it works fine. Can you share a zip of a reproduction?

**Comment by jaekschmitt** [comment]
@EvHaus did you ever get this figured out. I'm running into the exact same problem.

**Comment by EvHaus** [comment]
Nah, still broken for me. I'm using Yarn v1, which might be the issue. When I tried with pnpm and it worked fine.

@aklinker1 Here's a ZIP with a tiny monorepo where you can repro.

[wxt-monorepo.zip](https://github.com/wxt-dev/wxt/files/14423226/wxt-monorepo.zip)

To repro the issue, unzip it, then `cd wxt-monorepo` and run `yarn install` in the root of the repo. You'll get:

```
error /Users/evhaus/Git/wxt-monorepo/node_modules/wxt-starter: Command failed.
Exit code: 1
Command: wxt prepare
Arguments:
Directory: /Users/evhaus/Git/wxt-monorepo/node_modules/wxt-starter
Output:
WXT 0.16.11
ℹ Generating types...

 ERROR  Cannot find module '/Users/evhaus/Git/wxt-monorepo/packages/myextension/node_modules/wxt/dist/virtual/mock-browser.js'
Require stack:
- /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js

  Require stack:
  - /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js
  at Module._resolveFilename (node:internal/modules/cjs/loader:1072:15)
  at Function.resolve (node:internal/modules/helpers:127:19)
  at _resolve (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:251148)
  at jiti (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:253746)
  at /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/chunk-FNTE2L27.js:33:44
  at evalModule (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:254371)
  at /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/browser.js:23:29
  at evalModule (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (/Users/evhaus/Git/wxt-monorepo/node_modules/jiti/dist/jiti.js:1:254371)

✖ Command failed after 401 ms

 ERROR  Failed to load entrypoint: entrypoints/background.ts

  at importEntrypointFile (/Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:2051:13)
  at async getBackgroundEntrypoint (/Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:483:27)
  at async /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:287:18
  at async Promise.all (index 0)
  at async findEntrypoints (/Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:275:23)
  at async prepare (/Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:3472:23)
  at async /Users/evhaus/Git/wxt-monorepo/node_modules/wxt/dist/cli.js:3638:5
```

**Comment by aklinker1** [comment]
Interesting, so yarn doesn't create node_modules folders in each package? Only one in the root directory?

I'll try out your repro soon, thanks for sharing @EvHaus!

**Comment by aklinker1** [comment]
Here's the locations where we assume `node_modules` is in the WXT project's root. All these need to be changed to properly resolve to the correct node_modules folder.

https://github.com/wxt-dev/wxt/blob/50e8c86ce993347425b953bfd472f36e1f15e90b/src/core/builders/vite/plugins/virtualEntrypoint.ts#L32-L38

https://github.com/wxt-dev/wxt/blob/50e8c86ce993347425b953bfd472f36e1f15e90b/src/core/builders/vite/plugins/devHtmlPrerender.ts#L18-L21

https://github.com/wxt-dev/wxt/blob/50e8c86ce993347425b953bfd472f36e1f15e90b/src/core/builders/vite/plugins/webextensionPolyfillMock.ts#L28-L31

https://github.com/wxt-dev/wxt/blob/50e8c86ce993347425b953bfd472f36e1f15e90b/src/core/utils/building/import-entrypoint.ts#L54-L57

Unfortunately, I don't know how to properly resolve a package's absolute directory. If anyone has ideas, please share.

**Comment by aklinker1** [comment]
Workaround for now is to list `**/wxt` under the `nohoist` flag - this will tell yarn to put the module inside the package's `node_modules` directory, instead of the root one.

From @EvHaus 's repro, just add:

```diff
{
  "private": true,
  "name": "wxt-monorepo",
  "license": "UNLICENSED",
  "workspaces": {
    "packages": [
      "packages/*"
    ],
+   "nohoist": [
+     "**/wxt"
+   ]
  }
}
```

And everything works.

If someone knows how to resolve a node module's absolute path, I'll fix it. Otherwise use this workaround

**Comment by andrefgneves** [comment]
> If someone knows how to resolve a node module's absolute path, I'll fix it. Otherwise use this workaround

@aklinker1, I think you can do `require.resolve('wxt')` to get the full path to the nearest `wxt` "installation". You'd then use `path.resolve` to get the virtual modules paths.

**Comment by aklinker1** [comment]
> Edit: the nohoist "solution" doesn't work for me, unfortunately, as I'm using npm (with turbo) which does not support that "feature".

Sorry, repro was for yarn, I've never used NPM workspaces directly before.

> I think you can do require.resolve('wxt') to get the full path to the nearest wxt "installation". You'd then use path.resolve to get the virtual modules paths.

I'll give it a go, thanks!

**Comment by aklinker1** [comment]
@andrefgneves Could you try `wxt@0.17.4-alpha1`?

```sh
pnpm i wxt@next
```

**Comment by andrefgneves** [comment]
> @andrefgneves Could you try `wxt@0.17.4-alpha1`?

Yup, works perfectly! ❤️

**Comment by aklinker1** [comment]
Alright, I'll merge and release. Let you all know when it's available.

**Comment by aklinker1** [comment]
Released in [`v0.17.4`](https://github.com/wxt-dev/wxt/releases/tag/v0.17.4)
